transaction/summary/src/report.rs

```diff
@@ -3,90 +3,342 @@
 //! A TxSummaryUnblindingReport, containing the set of verified information
 //! about a transaction.
 
-use super::Error;
 use core::fmt::Display;
+
 use displaydoc::Display;
+use heapless::Vec;
 
 use mc_core::account::ShortAddressHash;
 use mc_transaction_types::{
     constants::{MAX_INPUTS, MAX_OUTPUTS},
     Amount, TokenId,
 };
-use mc_util_vec_map::VecMap;
+
+use super::Error;
 
 /// An entity with whom a transaction can interact, and who can be identified
 /// by the TxSummary verification process
 #[derive(Clone, Debug, Display, Eq, Ord, PartialEq, PartialOrd)]
 pub enum TransactionEntity {
-    /// Self
-    Ourself,
-    /// Address hash {0}
-    Address(ShortAddressHash),
+    /// Outputs to a non-change address that we control (hash {0})
+    OurAddress(ShortAddressHash),
+
+    /// Outputs to other accounts (hash {0})
+    OtherAddress(ShortAddressHash),
+
     /// Swap counterparty
     Swap,
 }
 
+/// Generic transaction report interface
+// (There is at this time only one report implementation, however, this trait
+// is particularly useful for eliding generics when using this and is expected
+// to be helpful when building support for account info caching.)
+pub trait TransactionReport {
+    /// Add value to the running transaction totals
+    fn input_add(&mut self, amount: Amount) -> Result<(), Error>;
+
+    /// Subtract an amount from the transaction total, used for change outputs
+    /// and SCIs if enabled
+    fn change_sub(&mut self, amount: Amount) -> Result<(), Error>;
+
+    /// Add SCI input not owned by our account
+    fn sci_add(&mut self, _amount: Amount) -> Result<(), Error>;
+
+    /// Add output value for a particular entity / address to the report
+    fn output_add(&mut self, entity: TransactionEntity, amount: Amount) -> Result<(), Error>;
+
+    /// Set the network fee
+    fn network_fee_set(&mut self, amount: Amount) -> Result<(), Error>;
+
+    /// Set the tombstone block
+    fn tombstone_block_set(&mut self, value: u64) -> Result<(), Error>;
+
+    /// Finalise the report, checking balances and sorting report entries
+    fn finalize(&mut self) -> Result<(), Error>;
+}
+
+/// [TransactionReport] impl for `&mut T` where `T: TransactionReport`
+impl<T: TransactionReport> TransactionReport for &mut T {
```

Comment:
> if T: TransactionReport doesn't that already imply &mut T is as well?

in some cases for calling but, not completely! if you want to be able to hold a reference to a type implementing a trait rather than an owned object you have to also use blanket forwarding impls like [this](https://doc.rust-lang.org/src/std/io/impls.rs.html#17).